File: node/src/application/useCase/CupomDescontoValidador.ts
import CupomDesconto from "../../domain/CupomDesconto";
import CupomRepositorio from "../../domain/CupomRepositorio";
import CupomDTO from "../dto/CupomDTO";
import ResultadoCupomValidador from "../dto/ResultadoCupomValidador";

export default class CupomDescontoValidador{

    private repositorio: CupomRepositorio;


    constructor(repositorio: CupomRepositorio){
        this.repositorio = repositorio;
    }

    public ehValido(cupomDTO: CupomDTO, data: Date): ResultadoCupomValidador{
        let resultado = false;
        const cupom = this.repositorio.getCupom(cupomDTO.codigoCupom);
        if (cupom){
            resultado = !cupom.estaExpirado(data);
        }
        return new ResultadoCupomValidador(resultado);
    }
}

File: golang/src/userCase/cupomDescontoValidador.go
package userCase

import (
	"clean-arq-go/src/domain"
	"clean-arq-go/src/userCase/DTOs"
)

type DescontoValidador struct {
	Repositorio domain.CupomRepositorio
}

func (validador DescontoValidador) EhValido(cupomDTO DTOs.CupomDTO) DTOs.ResultadoCupomValidador {
	cupom := validador.Repositorio.GetCupom(cupomDTO.Codigo)
	if cupom != nil {
		expirado := cupom.CupomExpirado()
		if !expirado {
			return DTOs.ResultadoCupomValidador{Messagem: "Cupom é valido", EhValido: true}
		}
	}
	return DTOs.ResultadoCupomValidador{Messagem: "Cupom é invalido", EhValido: false}
}
